### PR TITLE
vprof: init at 0.38

### PR DESCRIPTION
**Pull request by GaetanLepage**
## Description of changes

Add [vprof](https://github.com/nvdv/vprof), a visual profiler for Python.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### Priorities

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 271419` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python310Packages.vprof</li>
    <li>python310Packages.vprof.dist</li>
    <li>vprof (python311Packages.vprof)</li>
    <li>vprof.dist (python311Packages.vprof.dist)</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 271419` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python310Packages.vprof</li>
    <li>python310Packages.vprof.dist</li>
    <li>vprof (python311Packages.vprof)</li>
    <li>vprof.dist (python311Packages.vprof.dist)</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 271419` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python310Packages.vprof</li>
    <li>python310Packages.vprof.dist</li>
    <li>vprof (python311Packages.vprof)</li>
    <li>vprof.dist (python311Packages.vprof.dist)</li>
  </ul>
</details>

**Comment by natsukium** [comment]
Don't we need node-dependencies for UI?

**Comment by GaetanLepage** [comment]
You are right. Should I declare the node package in a `let` `in` block ?

**Comment by GaetanLepage** [comment]
@natsukium After diving in I was able to correctly vendor the necessary javascript file.
I tested locally and it works fine !

File: pkgs/development/python-modules/vprof/default.nix [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    patches = [
+      # buildNpmPackage would complain about the package not having a version.
+      # So we add one manually
+      ./add-version-to-ui-npm-package.patch
+    ];
+
+    postPatch = ''
+      # See comment on the patch above
+      substituteInPlace package.json \
+        --replace "@version@" "${version}"
```
Comment (GaetanLepage) [inline pkgs/development/python-modules/vprof/default.nix:35]: Actually, I could do a simpler thing: hardcode the version to be `1.0` in the patch.
No need to add this useless substitution afterwards.
The version of the ui dependency has absolutely no impact.

**Comment by natsukium** [comment]
Great. But we don't prefer to commit a lock file because it would make the nixpkgs tarball too heavy.
The best way is to ask upstream to add package-lock.json, but for now it might be better to get the source from PyPI.

**Comment by natsukium** [comment]
No, unfortunately, we have no choice.😢

**Review by natsukium — APPROVED** [review]
It works fine on aarch64-darwin.
LGTM

**Pull request merged**